src/create-with-edge-spec.ts

```diff
@@ -20,70 +20,76 @@ import { ResponseTypeToContext } from "./types/context.js"
-        }),
-        // this serializes responses that are returned by the route function,
-        // validating them against the routeSpec
-        serializeResponse(globalSpec, routeSpec),
-      ],
-      routeFn,
-      request,
-      ctx
-    )
+  return (routeSpec) => (routeFn) => {
+    const requestHandler: EdgeSpecRouteFn = async (request, ctx) => {
+      const onMultipleAuthMiddlewareFailures =
+        globalSpec.onMultipleAuthMiddlewareFailures ??
+        routeSpec.onMultipleAuthMiddlewareFailures
+
+      const supportedAuthMiddlewares = new Set<string>(
+        routeSpec.auth == null || routeSpec.auth === "none"
+          ? []
+          : Array.isArray(routeSpec.auth)
+            ? routeSpec.auth
+            : [routeSpec.auth]
+      )
+
+      const authMiddlewares = Object.entries(globalSpec.authMiddleware)
+        .filter(([k, _]) => supportedAuthMiddlewares.has(k))
+        .map(([_, v]) => v)
+
+      return await wrapMiddlewares(
+        [
+          // Injected into the VM when running in WinterCG emulation mode
+          // @ts-expect-error
+          ...(typeof _injectedEdgeSpecMiddleware !== "undefined"
+            ? // @ts-expect-error
+              _injectedEdgeSpecMiddleware
+            : []),
+          withUnhandledExceptionHandling,
+          // this serializes responses that are returned by middleware WITHOUT
+          // validating them against the routeSpec
+          //
+          // this allows returning EdgeSpecResponse.json or ctx.json in a
+          // middleware, instead of having to return a raw Response
+          //
+          // this is needed, for instance, when an error middleware returns an
+          // error response that does not match the routeSpec's response shape
+          serializeResponse(globalSpec, routeSpec, false),
+          ...(globalSpec.beforeAuthMiddleware ?? []),
+          firstAuthMiddlewareThatSucceeds(
+            authMiddlewares,
+            onMultipleAuthMiddlewareFailures
+          ),
+          ...(globalSpec.afterAuthMiddleware ?? []),
+          ...(routeSpec.middleware ?? []),
+          withMethods(routeSpec.methods),
+          withInputValidation({
+            supportedArrayFormats: globalSpec.supportedArrayFormats ?? [
+              "brackets",
+              "comma",
+              "repeat",
+            ],
+            commonParams: routeSpec.commonParams,
+            formData: routeSpec.multiPartFormData,
+            jsonBody: routeSpec.jsonBody,
+            queryParams: routeSpec.queryParams,
+            routeParams: routeSpec.routeParams,
+            urlEncodedFormData: routeSpec.urlEncodedFormData,
+          }),
+          // this serializes responses that are returned by the route function,
+          // validating them against the routeSpec
+          serializeResponse(globalSpec, routeSpec),
+        ],
+        routeFn,
+        request,
+        ctx
+      )
+    }
+
+    requestHandler._routeSpec = routeSpec
```

Comment:
Can we convert this to a functional pattern? i.e. `withAttachRouteSpec(routeSpec, ...)`

I think that would be more readable

Reply:
yeah that makes sense

Reply:
484dc88